src/types/internal/schema-field.ts

```diff
@@ -0,0 +1,328 @@
+          break;
+        case Kind.Inline:
+        case Kind.TypeName:
+          newField = new InlineField(field.model);
+          break;
+        case Kind.Union:
+          newField = new UnionField(field.types.map(type => SchemaField.create(type)));
+          break;
+        case Kind.Tuple:
+          newField = new TupleField(field.types.map(type => SchemaField.create(type)));
+          break;
+        case Kind.Nested:
+          newField = new NestedField(SchemaField.create(field.schema));
+          break;
+        default:
+          throw new Error(`Unsupported schema field ${field.kind}`);
+      }
+    }
+    newField.refinement = field.refinement || null;
+    newField.annotations = field.annotations || [];
+    return newField;
+  }
+
+  // The implementation of fromLiteral creates a cyclic dependency, so it is
+  // separated out. This variable serves the purpose of an abstract static.
+  static fromLiteral: SchemaFieldMethod = null;
+}
+
+export class PrimitiveField extends SchemaField {
+  constructor(public readonly type: SchemaPrimitiveTypeValue) {
+    super(Kind.Primitive);
+    assert(this.type);
+  }
+
+  getType(): SchemaPrimitiveTypeValue { return this.type; }
+
+  toString(): string { return this.type; }
+
+  normalizeForHash(): string { return `${this.type}|`; }
+
+  // tslint:disable-next-line: no-any
+  toLiteral(): {} {
+    return {...super.toLiteral(), type: this.type};
+  }
+}
+
+export class KotlinPrimitiveField extends SchemaField {
+  constructor(public readonly type: KotlinPrimitiveTypeValue) {
+    super(Kind.KotlinPrimitive);
+  }
+  getType(): KotlinPrimitiveTypeValue { return this.type; }
+
+  toString(): string { return this.type; }
+
+  normalizeForHash(): string { return `${this.type}|`; }
+
+  // tslint:disable-next-line: no-any
+  toLiteral(): {} {
+    return {...super.toLiteral(), type: this.type};
+  }
+}
+
+export class CollectionField extends SchemaField {
+  constructor(public readonly schema: SchemaField) {
+    super(Kind.Collection);
+  }
+
+  getSchema(): SchemaField { return this.schema; }
+
+  getModel(): EntityType {
+    return this.getSchema().getSchema() ? this.getSchema().getSchema().getModel() : null;
+  }
+
+  toString(): string { return `[${this.schema.toString()}]`; }
+
+  normalizeForHash(): string {
+    if (this.schema.isPrimitive || this.schema.isKotlinPrimitive) {
+      return `[${this.schema.getType()}]`;
+    }
+    return `[${this.schema.normalizeForHash()}]`;
+  }
+
+  isAtLeastAsSpecificAs(other: SchemaField): boolean {
+    assert(this.kind === other.kind);
+    return this.getSchema().isAtLeastAsSpecificAs(other.getSchema());
+  }
+
+  // tslint:disable-next-line: no-any
+  toLiteral(): {} {
+    return {kind: this.kind, schema: this.schema.toLiteral()};
+  }
+}
+
+export class ReferenceField extends SchemaField {
+  constructor(public readonly schema: SchemaField) {
+    super(Kind.Reference);
+  }
+  getSchema(): SchemaField { return this.schema; }
```

Comment:
These aren't really schemas; they're SchemaFields; or actually, schema types. They'll have different names (probably) depending on the field type - e.g. this might be getReferencedType.

Reply:
renamed to `getFieldType`, as it returns a `FieldType` object, and I didn't want to overload "reference" even more.
also renamed `getTypes` to `getFieldTypes` for consistency.

I didn't (yet) rename the actual data members, only the getter methods, because they are still used directly in some places. There will be more cleanups following this PR, i just prefer to make them incremental.

Reply:
Sounds good.